### PR TITLE
fix: use lockfile to protect self-increment of project id

### PR DESCRIPTION
**Pull request by gufeijun**

### CODE REVIEW

File: pkg/provisioner/helper_hostpath.go [hunk 1]
```diff
@@ -241,71 +252,89 @@ func (pOpts *HelperPodOptions) validateLimits() error {
...
 	// Extract the base path and the volume unique path.
-	var vErr error
-	config.parentDir, config.volumeDir, vErr = hostpath.NewBuilder().WithPath(pOpts.path).
+	config.parentDir, config.volumeDir, err = hostpath.NewBuilder().WithPath(pOpts.path).
 		WithCheckf(hostpath.IsNonRoot(), "volume directory {%v} should not be under root directory", pOpts.path).
+		WithCheckf(notReservedVolumeDir(), "volume directory '%s' is a reserved name", config.volumeDir).
 		ExtractSubPath()
-	if vErr != nil {
-		return vErr
+	if err != nil {
```
Comment (cnut) [inline pkg/provisioner/helper_hostpath.go:269]: is changing the vErr/lErr necessary?

File: pkg/provisioner/helper_hostpath.go [hunk 2]
```diff
@@ -241,71 +252,89 @@ func (pOpts *HelperPodOptions) validateLimits() error {
...
+		ProjectPath:         filepath.Join("/data/", config.volumeDir),
+		SoftLimitGrace:      config.pOpts.softLimitGrace,
+		UpperSoftLimitGrace: strings.ToUpper(config.pOpts.softLimitGrace),
+		HardLimitGrace:      config.pOpts.hardLimitGrace,
+		UpperHardLimitGrace: strings.ToUpper(config.pOpts.hardLimitGrace),
+	}); err != nil {
 		return err
 	}
+	lockfile := filepath.Join("/data/", lockfileNameForProjectID)
+	config.pOpts.cmdsForPath = []string{"flock", lockfile, "-c", strBuilder.String()}
```
Comment (zjx20) [inline pkg/provisioner/helper_hostpath.go:331]: Add a 30s timeout instead of waiting forever.

File: pkg/provisioner/helper_hostpath.go (hunk 1 above)
Comment (gufeijun) [reply pkg/provisioner/helper_hostpath.go:269]: Not necessary, but using single err variable is the idiomatic practice in go.
Example: [Go std](https://github.com/golang/go/blob/release-branch.go1.20/src/net/http/request.go#L555)

Comment (zjx20) [reply pkg/provisioner/helper_hostpath.go:269]: Yes, we should follow the practice when writing new code. But for the code ported from elsewhere, we should leave it as unchanged as possible.

**Review by cnut — APPROVED** [review]

**Review by zjx20 — APPROVED** [review]
Please update the title of this PR.

**Pull request merged**